// dynamodb-operations/index.js
'use strict'
const { DynamoDB } = require('aws-sdk')
const client = new DynamoDB.DocumentClient()

class Dynamo {

    constructor(table) {
        this.table = table
    }
  
    async insert(item) {
        const params = {
            TableName: this.table,
            Item: item
        }
        return await execute(client.put.bind(client), params)
    }

    async scan(params) {
        params.TableName = this.table
        return await execute(client.scan.bind(client), params)
    }

    async query(params) {
        params.TableName = this.table
        return await execute(client.query.bind(client), params)
    }

    async update(params) {
        params.TableName = this.table
        return await execute(client.update.bind(client), params)
    }

    async get(params) {
        params.TableName = this.table
        return await execute(client.get.bind(client), params)
    }
}

/**
 * Executa uma operação no dynamodb
 * @param {Function} method O method que será executado
 * @param {Object} params Os parâmetros do método
 */
async function execute(method, params) {
    return new Promise((resolve, reject) => {
        method(params, (err, data) => {
            if (err)
                reject(err)
            else
                resolve(data)
        })
    })
}

module.exports = Dynamo
